Dynamic Inputs

numso opened the pull request:
fixes https://github.com/chrisjpatty/flume/issues/32

Hey! I'm working on allowing nodes to define dynamic inputs but I wanted to post something here and get some feedback before I went any further. I'm open to any and all suggestions on direction/naming/etc!

This PR enables you to define `dynamicInputs` when you register a NodeType. This new config must be a function with the signature: `(InputData, PortBuilders) => [Ports]`. It gets called whenever you update your node's data using Controls (we're not resolving connections). Here's an example of how you might use it:

```
.addNodeType({
    type: 'compose',
    label: 'Compose',
    initialWidth: 250,
    inputs: ports => [
      ports.text({ name: "template", label: "Template", hidePort: true }),
    ],
    outputs: ports => [
      ports.text({ name: "message", label: "Message" }),
    ],
    dynamicInputs: (data, ports) => {
      const re = /\{(.*?)\}/g
      const ids = [...data.template.text.matchAll(re)].map(([,id]) => id)
      return [...new Set(ids).values()]
        .map(id => ports.text({ name: id, label: id }))
    }
  })
```

![image](https://user-images.githubusercontent.com/1761757/90711935-29d7cb00-e25f-11ea-9017-8937318b2941.png)

Is this the right direction? Is `dynamicInputs` the right name? Once we're ready I can take a shot at writing some docs and tests.

I'm curious about the line I commented out in Control.js (a call to `triggerRecalculation()`). Is this call still needed? It looks like `triggerRecalculation` is passed into the `"custom"` controls to be invoked when needed. And it doesn't seem to do anything when called after updating the normal controls. It ends up being called on every keypress in a text control which, aside from being inefficient, busted this feature: Things ran a bit out of order and flume tried to recalculate the positions of ports that no longer existed. If this is still needed then one alternative is to update `getPortRect` in `connectionCalculator.js` to handle missing DOM nodes.

numso commented:
I just realized this would also allow you to make arrays 😛 

```
    inputs: ports => [
      ports.number({ name: "length", label: "Length", hidePort: true })
    ],
    dynamicInputs: (data, ports) => {
      const length = data.length.num || 0
      return [...new Array(length)]
        .map((_, i) => ports.text({ name: `item${i}`, label: `Item ${i}` }))
    }
```
![image](https://user-images.githubusercontent.com/1761757/90714019-ef246180-e263-11ea-81e4-96bbc9efc493.png)

Its not perfect but it's usable until flume officially supports them.

chrisjpatty commented:
Thanks @numso! This is really great work. I really like your example for how this would be useful, it makes the value-add really clear. Looking through this proposed API I have a few thoughts:
- Having a separate `dynamicInputs` key is idiomatic, but might have a couple constraints:
  - Dynamic inputs have to render below the static inputs, and can't be mixed in.
  - Technically all static inputs could also be defined in the dynamic inputs function, so there's some duplication of functionality.
- Inputs should also probably be allowed to depend on the context object as well as the current input data. That should be easy enough to forward to this builder function though.
- What if instead of a new key for dynamic inputs, what do you think about allowing a curried function for dynamic inputs. So for example, you could define your nodes like normal, or you could pass a function like this:

```js
.addNodeType({
  type: 'compose',
  label: 'Compose',
  initialWidth: 250,
  inputs: ports => (inputData, context) => {
    const re = /\{(.*?)\}/g
    const ids = [...data.template.text.matchAll(re)].map(([,id]) => id)
    return [
      ports.number({ name: "length", label: "Length", hidePort: true }),
      [...new Set(ids).values()]
      .map(id => ports.text({ name: id, label: id }))
    ]
  },
  outputs: ports => [
    ports.text({ name: "message", label: "Message" }),
  ],
})
```
Then we just check if the inputs are a function or an array, and handle it accordingly. This syntax is similar to the pattern that Redux Thunk uses so you have an opt-in strategy for adding additional functionality to the existing inputs key. This way the static ports can be intermixed with dynamic ones as well, which there may be some use-cases for. 

There's obviously more work that would need to be done to ensure that connections are properly cleaned up for inputs that are destroyed, but it looks like you've already laid some groundwork for that. Really great PR, let me know your thoughts on the above API.

numso commented:
I love that! Makes a ton of sense and is much cleaner than what I've got now. I'll make those changes in the next day or two.

numso commented:
Alright! I've updated the RootEngine to work with this feature and created some basic documentation describing dynamic inputs and outputs. I had to comment out the examples used in those new docs until we upgrade flume in the docs/ project as they use the new dynamic inputs/outputs feature.

I think we're getting close. I've been testing it out and everything works like I'd expect. I wanted to touch on a couple things:
1. There is a function in typeBuilders for removing a portType. It checks to make sure there are no nodes which use that portType as an input or output. We will no longer have a way of validating that as inputs/outputs can now be defined as functions:
https://github.com/chrisjpatty/flume/blob/680a58c89ead452f2194c0b2e9c39f9735fa4b9e/src/typeBuilders.js#L268-L275
2. When a node is created, we give it some default inputData based on it's inputs. As values change that node may now gain or lose inputs. When this happens we do not generate new defaults.
https://github.com/chrisjpatty/flume/blob/680a58c89ead452f2194c0b2e9c39f9735fa4b9e/src/nodesReducer.js#L284-L287

I must not be thinking of the right use cases because I don't understand why someone would want to remove a port type. So I'm not sure if item 1 is acceptable or not. As for item 2, if we leave it as is then the resolvePorts documentation may need to change a bit. Right now it says something like this:
```
const resolvePorts = (portType, data) => {
  switch (portType) {
    case 'string':
      return data.string
...
```
Without a guarantee of defaults, `data` might be undefined and would need to be checked: `data?.string ?? ''`. Ideally we'd keep the defaults up to date and this wouldn't be a problem. I'll try a couple things out tomorrow and see if we can't find a performant way to do that.

chrisjpatty commented:
Fantastic work! I'm going to take a closer look at this tomorrow, but I've got some quick thoughts as I read through your comments. 
1. The remove functions in the type builders are useful in some exotic cases, but are currently undocumented. I'm using them in a project where I start with a base config with a bunch of available ports and nodes, and I then have files which remove some of the ports and nodes for sub logic editors which don't support them. The same thing could be accomplished by structuring my files differently so it's not necessarily required functionality, just semantically convenient. I'd be ok with doing a cursory check in this function the way it does, and if it encounters at least one function it just throws a warning that when removing types when using dynamic ports, the developer must ensure that they are not used in any node types. (This could also just be in the documentation if we don't want to throw warnings). It's a bit of an exotic use case so I'm okay with this function existing with some potential pitfalls as long as they're relatively documented. 
2. This is the trickier one, and it's the primary sticking point with dynamic ports in the first place. I might need to think about this more tomorrow, but if we do update the default data, we can't immediately react to the function being called with different ports or we'd introduce side-effects during the render phase (which won't be concurrent-mode compatible). So we'd probably have to mark in the state somewhere that the default data needs to be updated and clean it up in a `useEffect`, but that doesn't solve the problem of the data being out of sync for the initial render. I'm sort of rambling a bit, but I think ideally we'll need to reconcile the old input data with the new input data somehow. Dynamic inputs are definitely an advanced use-case so I'm ok with some of the onus being on the end-developer to maintain their own assumptions in the engine resolution functions, and it being documented as such, but as much as we can abstract away to improve the developer-experience the better. 

Again, great work! Thanks so much for taking this feature on, it'll be a huge leap forward for the library once it's integrated. I'll try to dive a little deeper into some of the changes tomorrow, but it's looking great so far.

chrisjpatty commented:
BTW, I'm working on trying to resolve the issue with the docs not inheriting the local version of Flume. Ideally I should be able to get it linked such that it builds the local flume, and then builds the docs using the dist folder. 🤞

numso commented:
Ah, your explanation for why you'd use `removePortType` makes total sense! That may actually clean up some of my configs :). I'll go add that warning right now.

chrisjpatty commented:
Apologies that it's taken me a minute to get to reviewing this PR, I ended up getting swamped at work last week. I've pulled it down and run some tests and I've got some feedback I'm writing up, but overall this is amazing work! I think just a couple of tweaks and it will be ready to go. Thanks again for your great work on this.

chrisjpatty reviewed (commented):
@numso Everything you have in these changes look great! I've created some of my own test case nodes to test out this feature and for the most part it worked really well, especially since I knew the constraints ahead of time. My small concern is what we've already talked about, that this feature breaks some of the built-in type safety features without some extra work on the end-developer's side, namely the problem of input data not matching the schema of the dynamic ports. I have a couple thoughts about some ways we could smooth over that part, and I wanted to get your thoughts before we merge this feature in. 

One thought I had was to allow the end-developer to supply an optional reconciliation function for reconciling input data between different port schemas. That way you can handle that complexity in the type builders instead of polluting the root engine with a bunch of type checking that wouldn't otherwise be necessary. 

Here's an example of what I'm thinking:

```js
config
  .addPortType({
    type: "string",
    name: "string",
    label: "text",
    controls: [Controls.text()]
  })
  .addPortType({
    type: "animals",
    name: "animals",
    label: "Animals",
    controls: [
      Controls.multiselect({
        label: "Animals",
        options: [
          { value: "Dog", label: "Dog" },
          { value: "Cat", label: "Cat" },
          { value: "Horse", label: "Horse" },
          { value: "Chicken", label: "Chicken" },
          { value: "Cow", label: "Cow" }
        ]
      })
    ]
  })
  .addNodeType({
    type: "animalNamer",
    label: "Animal Namer",
    description: "Outputs a provided name for each selected animal.",
    inputs: ports => (inputData, connections, context, reconcileInputData) => {
      const selectedAnimals = inputData.animals
        ? inputData.animals.animals
        : [];

      reconcileInputData(() => {
        let newData = {
          animals: inputData.animals
        };
        selectedAnimals.forEach(animal => {
          newData[animal] = {
            text: inputData[animal] ? inputData[animal].text : ""
          };
        });
        return newData;
      });

      return [
        ports.animals(),
        ...selectedAnimals.map(animal =>
          ports.text({ name: animal, label: `${animal} Name` })
        )
      ];
    },
    outputs: ports => inputData => {
      const selectedAnimals = inputData.animals
        ? inputData.animals.animals
        : [];
      return selectedAnimals.map(animal =>
        ports.text({ name: `${animal}Name`, label: `${animal} Name` })
      );
    }
  });
```

Hopefully this example isn't too contrived to be useful. Basically this node lets you pick animals from a list, and then creates an input and output text node for each animal for giving each animal a name. In this example the ports function provides a function called `reconcileInputData`, which expects to be called with a function that returns a new set of inputData. On the library side this function is called called, and the node is updated with the new data:

```js
// IoPorts.js

function useTransputs(
  transputsFn,
  transputType,
  nodeId,
  inputData,
  connections
) {
  const nodesDispatch = React.useContext(NodeDispatchContext);
  const executionContext = React.useContext(ContextContext);
  const reconcileInputData = React.useRef()

  const transputs = React.useMemo(() => {
    if (Array.isArray(transputsFn)) return transputsFn;
    return transputsFn(inputData, connections, executionContext, reconcile => {
      reconcileInputData.current = reconcile
    });
  }, [transputsFn, inputData, connections, executionContext, reconcileInputData]);
  const prevTransputs = usePrevious(transputs);

  React.useEffect(() => {
    if (!prevTransputs || Array.isArray(transputsFn)) return;
    if(typeof reconcileInputData.current === "function"){
      nodesDispatch({
        type: "SET_NODE_INPUT_DATA",
        inputdata: reconcileInputData.current()
      })
    }
    for (const transput of prevTransputs) {
      const current = transputs.find(({ name }) => transput.name === name);
      if (!current) {
        nodesDispatch({
          type: "DESTROY_TRANSPUT",
          transputType,
          transput: { nodeId, portName: transput.name }
        });
      }
    }
  }, [
    transputsFn,
    transputs,
    prevTransputs,
    nodesDispatch,
    nodeId,
    transputType
  ]);

  return transputs;
}
```
(This is just quickly written code for example's sake, it may not even be working correctly I'm not sure). Basically the idea is to cache the developer's reconcile function, and then call it only when potentially necessary. I'm not sure that I'm actually in love with this pattern though, it feels a little awkward. It's possible we could maybe even make this an automatic resolution process that compares the old and new ports similar to how your `useTransputs` function is already doing, and then automatically reconcile the input data. You've already setup memoization in here so it may be reasonably performant to do that, but it would need to be tested. 

Anyways, I just wanted to get your thoughts on this. I think that this is probably outside of the scope of your PR. What have setup is working great and I'd be happy to merge it in after resolving the small attached change in the root engine. That way  people who need this feature can start using it right away, and we can work on improving the developer experience in a later PR, but I just wanted to get your thoughts on how we could smooth over some of the harder parts of using dynamic transputs.

Thanks again for working on this feature, it's a big leap forward for the library!

chrisjpatty commented:
@numso BTW I just merged in a fix for the docs. You theoretically should be able to merge master into this branch now and then you can uncomment your documentation examples without needing to use a published version of these changes.

numso commented on src/RootEngine.js:
```diff
@@ -44,7 +44,11 @@ class RootEngine {
+    if (typeof inputs === 'function') {
+      inputs = inputs(node.inputData, node.connections, context)
+    }
+    return inputs.reduce((obj, input) => {
```
Makes total sense; I'm on it!

numso commented:
Thanks for the comments! I think you are right that we should address the de-synced inputData in a future PR. I don't think we've had enough time to think about and come up with a good API for that case. I'll create a separate issue to track and discuss that. We should also probably file an issue for array ports so we don't lose track of that discussion when this PR is merged; I can do that too.

> I just merged in a fix for the docs.

Awesome news! I'll uncomment the docs in this PR and make sure they're all still up to date.

Just FYI: While testing out this feature myself I ran into two more small issues that I want to fix before we merge. And I think it'd be wise to write a handful of tests to cover this new functionality. I believe I'll have time later today to make all those changes.

numso commented:
Alright, I just added some basic tests. I also fixed those last two issues and wrote tests for them in the same commit.

Looks like netlify is failing now because I'm using `String.prototype.matchAll` in one of my nodeTypes but netlify is building with node v10.22.0 which [doesn't support that](https://node.green/#ES2020-features-String-prototype-matchAll-basic-functionality). I believe node 10 is no longer the active LTS and is now in maintenance. Should we have netlify build with a later version of node or do you want me to refactor that code to be node v10 compatible?

chrisjpatty commented:
@numso Sorry for the late reply, I can change the node version to 12.x if we need I think. I think Netlify lets us just set the `NODE_VERSION` variable. Or does your latest commit make that unnecessary? It looks like it's building now.

numso commented:
No worries. That is correct: we do not need to update the node version for this PR. My last commit makes the docs compatible with node 10 and everything is working.

I have been using this branch in my project without issue for the past couple weeks. I believe this feature is ready to go in. Let me know if you find any more issues or think of any changes that should be made.

chrisjpatty commented:
Ok enough procrastinating on my end, I'm going to get this merged in and published as version `0.6.0`. Thanks again for your work on this, I already have ideas for how to use it in my own work!
